sycl/source/detail/plugin.hpp

```diff
@@ -25,7 +25,67 @@ namespace sycl {
 namespace detail {
 #ifdef XPTI_ENABLE_INSTRUMENTATION
 extern xpti::trace_event_data_t *GPICallEvent;
+extern xpti::trace_event_data_t *GPIArgCallEvent;
 #endif
+
+template <PiApiKind Kind, size_t Idx, typename... Args>
+struct array_fill_helper;
+
+template <PiApiKind Kind> struct PiApiArgTuple;
+
+#define _PI_API(api, ...)                                                      \
+  template <> struct PiApiArgTuple<PiApiKind::api> {                           \
+    using type = std::tuple<__VA_ARGS__>;                                      \
+  };
+
+#include <CL/sycl/detail/pi.def>
+#undef _PI_API
+
+template <PiApiKind Kind, size_t Idx, typename T>
+struct array_fill_helper<Kind, Idx, T> {
+  static void fill(unsigned char *Dst, size_t Offset, T &&Arg) {
+    using ArgsTuple = typename PiApiArgTuple<Kind>::type;
+    // C-style cast is required here.
+    auto RealArg = (typename std::tuple_element<Idx, ArgsTuple>::type)(Arg);
+    auto *Begin = reinterpret_cast<const unsigned char *>(&RealArg);
+    auto *End = Begin + sizeof(decltype(RealArg));
+    std::uninitialized_copy(Begin, End, Dst + Offset);
+  }
+};
+
+template <PiApiKind Kind, size_t Idx, typename T, typename... Args>
+struct array_fill_helper<Kind, Idx, T, Args...> {
+  static void fill(unsigned char *Dst, size_t Offset, const T &&Arg,
+                   Args &&... Rest) {
+    using ArgsTuple = typename PiApiArgTuple<Kind>::type;
+    // C-style cast is required here.
+    auto RealArg = (typename std::tuple_element<Idx, ArgsTuple>::type)(Arg);
+    auto *Begin = reinterpret_cast<const unsigned char *>(&RealArg);
+    auto *End = Begin + sizeof(decltype(RealArg));
+    std::uninitialized_copy(Begin, End, Dst + Offset);
```

Comment:
```suggestion
*(typename std::tuple_element<Idx, ArgsTuple>::type *)Dst = RealArg;
```
Assuming we get rid of offset arg(see next comment)